### PR TITLE
Oauth prompt updates

### PR DESCRIPTION
**Pull request by EricDahlvang**
Fixes https://github.com/microsoft/botbuilder-js/issues/2651 OAuthPrompt should provide the ability to cancel itself (JS)

Fixes https://github.com/microsoft/botbuilder-js/issues/2675 OAuthPrompt timeout applies to all auth related events and invokes.

### CODE REVIEW

File: libraries/botbuilder-dialogs/src/prompts/oauthPrompt.ts [hunk 1]
```diff
@@ -199,6 +215,8 @@ export class OAuthPrompt extends Dialog {
             // Return recognized value or re-prompt
             if (isValid) {
                 return await dc.endDialog(recognized.value);
+            } else if (isMessage && this.settings.endOnInvalidMessage) {
```
Comment (munozemilio) [inline libraries/botbuilder-dialogs/src/prompts/oauthPrompt.ts:218]: I always prefer to return early and use guard clauses in the code to keep it clean and easier to read, but some may argue having one entry/exit is a better practice.

Comment (EricDahlvang) [reply libraries/botbuilder-dialogs/src/prompts/oauthPrompt.ts]: changed, thanks @munozemilio

File: libraries/botbuilder-dialogs/src/prompts/oauthPrompt.ts [hunk 2]
```diff
@@ -55,6 +55,16 @@ export interface OAuthPromptSettings {
      * Defaults to a value `900,000` (15 minutes.)
      */
     timeout?: number;
+
+    /**
+     * (Opitonal) value indicating whether the OAuthPrompt should end upon
```
Comment (benbrown) [inline libraries/botbuilder-dialogs/src/prompts/oauthPrompt.ts:60]: typo here (Optional)

File: libraries/botbuilder-dialogs/tests/oauthPrompt.test.js [hunk 3]
```diff
@@ -749,3 +846,67 @@ function uuid() {
...
+    await adapter.send('Hello')
+        .assertReply(activity  => {
+            assert(activity.attachments.length === 1);
+            assert(activity.attachments[0].contentType === CardFactory.contentTypes.oauthCard);
+
+            adapter.addUserToken(connectionName, activity.channelId, activity.recipient.id, token, magicCode);
+            adapter.addExchangeableToken(connectionName, activity.channelId, activity.recipient.id, exchangeToken, token);
+        })  
+        .send(oauthPromptActivity)
+        .assertReply(noTokenResonse);
```
Comment (mdrichardson) [inline libraries/botbuilder-dialogs/tests/oauthPrompt.test.js:911]:
```suggestion
async function testTimeout(oauthPromptActivity, shouldSucceed = true, tokenResponse = 'Failed', noTokenResponse = 'Ended') {
    var connectionName = 'myConnection';
    var token = 'abc123';
    var magicCode = '888999';
    const exchangeToken = 'exch123';

    // Create new ConversationState with MemoryStorage
    const convoState = new ConversationState(new MemoryStorage());

    // Create a DialogState property, DialogSet and OAuthPrompt
    const dialogState = convoState.createProperty('dialogState');
    const dialogs = new DialogSet(dialogState);
    dialogs.add(new OAuthPrompt('prompt', {
        connectionName,
        title: 'Login',
        timeout: 0
    }, async (prompt) => {
        if (prompt.recognized.succeeded) {
            assert(shouldSucceed, 'recognition succeeded but should have failed during testTimeout');
            return true;
        }
        
        assert(!shouldSucceed, 'recognition failed during testTimeout');
        return false;
    }));

    // Initialize TestAdapter.
    const adapter = new TestAdapter(async (turnContext) => {
        const dc = await dialogs.createContext(turnContext);

        if(!oauthPromptActivity.conversation) {
            oauthPromptActivity.conversation = turnContext.activity.conversation;
            oauthPromptActivity.recipient = turnContext.activity.recipient;
            oauthPromptActivity.from = turnContext.activity.from;
            oauthPromptActivity.serviceUrl = turnContext.activity.serviceUrl;
            oauthPromptActivity.channelId = turnContext.activity.channelId;
        }

        const results = await dc.continueDialog();
        if (results.status === DialogTurnStatus.empty) {
            await dc.prompt('prompt', { });
        } else if (results.status === DialogTurnStatus.complete || (results.status === DialogTurnStatus.waiting && !shouldSucceed)) {
            if (results.result && results.result.token) {
                await turnContext.sendActivity(tokenResponse);
            }
            else {
                await turnContext.sendActivity(noTokenResponse);
            }
        }
        await convoState.saveChanges(turnContext);
    });

    await adapter.send('Hello')
        .assertReply(activity  => {
            assert(activity.attachments.length === 1);
            assert(activity.attachments[0].contentType === CardFactory.contentTypes.oauthCard);

            adapter.addUserToken(connectionName, activity.channelId, activity.recipient.id, token, magicCode);
            adapter.addExchangeableToken(connectionName, activity.channelId, activity.recipient.id, exchangeToken, token);
        })  
        .send(oauthPromptActivity)
        .assertReply(noTokenResponse);
```

typo

Comment (mdrichardson) [inline libraries/botbuilder-dialogs/tests/oauthPrompt.test.js:911]:
```suggestion
async function testTimeout(oauthPromptActivity, shouldSucceed = true, tokenResponse = 'Failed', noTokenResponse = 'Ended') {
    var connectionName = 'myConnection';
    var token = 'abc123';
    var magicCode = '888999';
    const exchangeToken = 'exch123';

    // Create new ConversationState with MemoryStorage
    const convoState = new ConversationState(new MemoryStorage());

    // Create a DialogState property, DialogSet and OAuthPrompt
    const dialogState = convoState.createProperty('dialogState');
    const dialogs = new DialogSet(dialogState);
    dialogs.add(new OAuthPrompt('prompt', {
        connectionName,
        title: 'Login',
        timeout: 0
    }, async (prompt) => {
        if (prompt.recognized.succeeded) {
            assert(shouldSucceed, 'recognition succeeded but should have failed during testTimeout');
            return true;
        }
        
        assert(!shouldSucceed, 'recognition failed during testTimeout');
        return false;
    }));

    // Initialize TestAdapter.
    const adapter = new TestAdapter(async (turnContext) => {
        const dc = await dialogs.createContext(turnContext);

        if(!oauthPromptActivity.conversation) {
            oauthPromptActivity.conversation = turnContext.activity.conversation;
            oauthPromptActivity.recipient = turnContext.activity.recipient;
            oauthPromptActivity.from = turnContext.activity.from;
            oauthPromptActivity.serviceUrl = turnContext.activity.serviceUrl;
            oauthPromptActivity.channelId = turnContext.activity.channelId;
        }

        const results = await dc.continueDialog();
        if (results.status === DialogTurnStatus.empty) {
            await dc.prompt('prompt', { });
        } else if (results.status === DialogTurnStatus.complete || (results.status === DialogTurnStatus.waiting && !shouldSucceed)) {
            if (results.result && results.result.token) {
                await turnContext.sendActivity(tokenResponse);
            }
            else {
                await turnContext.sendActivity(noTokenResponse);
            }
        }
        await convoState.saveChanges(turnContext);
    });

    await adapter.send('Hello')
        .assertReply(activity  => {
            assert(activity.attachments.length === 1);
            assert(activity.attachments[0].contentType === CardFactory.contentTypes.oauthCard);

            adapter.addUserToken(connectionName, activity.channelId, activity.recipient.id, token, magicCode);
            adapter.addExchangeableToken(connectionName, activity.channelId, activity.recipient.id, exchangeToken, token);
        })  
        .send(oauthPromptActivity)
        .assertReply(noTokenResponse);
```

typo

**Pull request merged**